File: src/main/java/io/github/kingg22/api/vacunas/panama/configuration/security/CustomJwtRefreshFilter.java
package io.github.kingg22.api.vacunas.panama.configuration.security;

import io.github.kingg22.api.vacunas.panama.service.ITokenService;
import jakarta.servlet.FilterChain;
import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import java.io.IOException;
import lombok.RequiredArgsConstructor;
import org.jetbrains.annotations.NotNull;
import org.springframework.security.oauth2.jwt.Jwt;
import org.springframework.security.oauth2.jwt.JwtDecoder;
import org.springframework.web.filter.OncePerRequestFilter;

@RequiredArgsConstructor
class CustomJwtRefreshFilter extends OncePerRequestFilter {
    private final ITokenService tokenService;
    private final JwtDecoder jwtDecoder;
    private static final String REFRESH_TOKEN_ENDPOINT = "/vacunacion/v1/token/refresh";

    /**
     * Filter to validate {@link Jwt} tokens on each request. This filter checks if the token present in the request is
     * a refresh token. If the request is made to the {@code REFRESH_TOKEN_ENDPOINT}, it allows access only if the
     * provided token is a valid refresh token. If a refresh token is used to access any other endpoint, it will be
     * rejected. Additionally, if an access token is used to access the {@code REFRESH_TOKEN_ENDPOINT} it will also be
     * rejected.
     *
     * @param request The {@link HttpServletRequest} containing the incoming request data.
     * @param response The {@link HttpServletResponse} used to send the response.
     * @param filterChain The {@link FilterChain} used to invoke the next filter in the chain.
     * @throws ServletException If an error occurs during the processing of the request.
     * @throws IOException If an input or output error occurs while handling the request.
     */
    @Override
    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain)
            throws ServletException, IOException {
        var authorization = request.getHeader("Authorization");

        if (authorization != null && authorization.startsWith("Bearer ")) {
            var token = authorization.substring("Bearer ".length());

            try {
                var jwt = jwtDecoder.decode(token);
                var userId = jwt.getSubject();
                var tokenId = jwt.getId();

                var accessTokenValid = tokenService.isAccessTokenValid(userId, tokenId);
                var refreshTokenValid = tokenService.isRefreshTokenValid(userId, tokenId);

                if (!accessTokenValid
                        && refreshTokenValid
                        && !request.getRequestURI().equals(REFRESH_TOKEN_ENDPOINT)) {
                    setWWWHeader(
                            response,
                            "invalid_token",
                            "An error occurred while attempting to decode the Jwt: Refresh token is only for refresh tokens");
                    response.setStatus(HttpServletResponse.SC_FORBIDDEN);
                    return;
                }
                if (accessTokenValid
                        && !refreshTokenValid
                        && request.getRequestURI().equals(REFRESH_TOKEN_ENDPOINT)) {
                    setWWWHeader(
                            response,
                            "invalid_token",
                            "An error occurred while attempting to decode the Jwt: Access token cannot be used to refresh tokens");
                    response.setStatus(HttpServletResponse.SC_FORBIDDEN);
                    return;
                }
                // TODO add validation if the type is false to both type
            } catch (RuntimeException exception) {
                setWWWHeader(
                        response,
                        "server_error",
                        "An error occurred while attempting to decode the Jwt: Token validation failed due to server issue");
                response.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
                return;
            }
        }
        filterChain.doFilter(request, response);
    }

    /**
     * Set header {@code WWW-Authenticate} based on the format that
     * {@link org.springframework.security.oauth2.core.OAuth2TokenValidatorResult} does for validation errors
     *
     * @param response The {@link HttpServletResponse} used to send the response.
     * @param errorCode The code based on {@link org.springframework.security.oauth2.core.OAuth2Error}
     * @param description The description of error based on {@link org.springframework.security.oauth2.core.OAuth2Error}
     * @see org.springframework.security.oauth2.core.OAuth2AuthenticationException
     */
    private void setWWWHeader(
            @NotNull final HttpServletResponse response,
            @NotNull final String errorCode,
            @NotNull final String description) {
        StringBuilder wwwAuthenticate = new StringBuilder();
        wwwAuthenticate.append("Bearer");
        if (!errorCode.isBlank() && !description.isBlank()) {
            wwwAuthenticate.append(" error=\"").append(errorCode).append("\", error_description=\"");
            wwwAuthenticate.append(description).append("\"");
        }
        response.setHeader("WWW-Authenticate", wwwAuthenticate.toString());
    }
}

File: src/main/kotlin/io/github/kingg22/api/vacunas/panama/configuration/security/CustomJwtRefreshFilter.kt
@file:Suppress("kotlin:S6508") // Inheritance / Implementation cannot change the signature

package io.github.kingg22.api.vacunas.panama.configuration.security

/*
TODO this is needed, but need adapt to quarkus
class CustomJwtRefreshFilter(private val tokenService: TokenService, private val jwtDecoder: ReactiveJwtDecoder) :
    WebFilter {
    companion object {
        private const val REFRESH_TOKEN_ENDPOINT = "/vacunacion/v1/token/refresh"
        private const val AUTHORIZATION_HEADER = "Authorization"
        private const val BEARER_PREFIX = "Bearer "

        private fun setWWWHeader(response: ServerHttpResponse, errorCode: String, description: String) {
            response.headers[HttpHeaders.WWW_AUTHENTICATE] = buildString {
                append("Bearer")
                if (errorCode.isNotBlank() && description.isNotBlank()) {
                    append(" error=\"$errorCode\", error_description=\"$description\"")
                }
            }
        }
    }

    private val log = logger()

    override fun filter(exchange: ServerWebExchange, chain: WebFilterChain): Mono<Void> {
        val authorization = exchange.request.headers.getFirst(AUTHORIZATION_HEADER)
        if (authorization.isNullOrBlank() || !authorization.startsWith(BEARER_PREFIX)) {
            return chain.filter(exchange)
        }

        val token = authorization.removePrefix(BEARER_PREFIX)
        log.debug("Trying to decode Token: {}", token)

        return jwtDecoder.decode(token)
            .flatMap { jwt ->
                val userId = jwt.subject
                val tokenId = jwt.id
                log.debug("Verifying token for $userId with tokenId: $tokenId")

                Mono.zip(
                    tokenService.isAccessTokenValid(userId, tokenId),
                    tokenService.isRefreshTokenValid(userId, tokenId),
                ).flatMap {
                    val (accessTokenValid, refreshTokenValid) = it.t1 to it.t2
                    val path = exchange.request.uri.path
                    checkNotNull(path) { "Path is null for filter chain" }
                    log.debug(
                        "IsAccessTokenValid: $accessTokenValid, IsRefreshTokenValid: $refreshTokenValid. Going to '$path'",
                    )

                    when {
                        !accessTokenValid &&
                            refreshTokenValid &&
                            path != REFRESH_TOKEN_ENDPOINT -> {
                            log.debug("Refresh token invalid use. Returning 403 Forbidden.")
                            setWWWHeader(exchange.response, "invalid_token", "Refresh token is only for refresh tokens")
                            exchange.response.statusCode = HttpStatus.FORBIDDEN
                            exchange.response.setComplete()
                        }

                        accessTokenValid &&
                            !refreshTokenValid &&
                            path == REFRESH_TOKEN_ENDPOINT -> {
                            log.debug("Access token invalid use. Returning 403 Forbidden.")
                            setWWWHeader(
                                exchange.response,
                                "invalid_token",
                                "Access token cannot be used to refresh tokens",
                            )
                            exchange.response.statusCode = HttpStatus.FORBIDDEN
                            exchange.response.setComplete()
                        }

                        !accessTokenValid && !refreshTokenValid -> {
                            log.debug("Access and refresh token invalid use. Returning 403 Forbidden.")
                            setWWWHeader(exchange.response, "invalid_token", "Tokens has been revoked")
                            exchange.response.statusCode = HttpStatus.FORBIDDEN
                            exchange.response.setComplete()
                        }

                        else -> chain.filter(exchange)
                    }
                }
            }
            .onErrorResume {
                log.error("Error occurred during JWT decoding in JwtRefreshFilter", it)
                setWWWHeader(exchange.response, "server_error", "Token validation failed due to server issue")
                exchange.response.statusCode = HttpStatus.INTERNAL_SERVER_ERROR
                exchange.response.setComplete()
            }
    }
}
*/
